### ISSUE TITLE
Reply from pod to outside is dropped when L7 ingress policy is used

### ISSUE THREAD

**Issue by brb** [cilium/cilium#21954]
This issue is about an L7 ingress policy problem a when a pod is reached directly from outside client / via a NodePort BPF service.

Let's consider the following L7 netpol:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: foobar
spec:
  description: "Allow to GET on echo from outside"
  endpointSelector:
    matchLabels:
      kind: echo
  ingress:
  - fromEntities:
    - "world"
    toPorts:
    - ports:
      - port: "8080"
        protocol: TCP
      rules:
        http:
        - method: "GET"
          path: "/$"
```

When the netpol is applied, accessing the `echo` pod from outside the cluster fails with:
```
xx drop (Stale or unroutable IP) flow 0xf863e34e to endpoint 0, file bpf_host.c line 665, , identity 28143->unknown: 10.0.1.49:80 -> 192.168.34.1:32884 tcp SYN, ACK
```
The drop is triggered by https://github.com/cilium/cilium/blob/v1.12.3/bpf/bpf_host.c#L626.

What happens is that the L7 proxy sends the SYN-ACK which gets handled by `bpf_host @ cilium_host`, and then dropped. See the  `pwru` output (ifindex=9 is `cilium_host`):

```
  SKB    CPU          PROCESS                     FUNC
0xffff9a859048e800      7        [<empty>]             ip_local_out netns=4026531992 mark=0xa00 ifindex=0 proto=0 mtu=0 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]           __ip_local_out netns=4026531992 mark=0xa00 ifindex=0 proto=0 mtu=0 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]             nf_hook_slow netns=4026531992 mark=0xa00 ifindex=0 proto=8 mtu=0 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]                ip_output netns=4026531992 mark=0xa00 ifindex=0 proto=8 mtu=0 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]             nf_hook_slow netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]  apparmor_ipv4_postroute netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]         ip_finish_output netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>] __cgroup_bpf_run_filter_skb netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]       __ip_finish_output netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]        ip_finish_output2 netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]           dev_queue_xmit netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]         __dev_queue_xmit netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]             tcf_classify netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]      skb_ensure_writable netns=4026531992 mark=0x0 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]                kfree_skb netns=4026531992 mark=0x0 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
```

The packet is sent to the `cilium_host` because of the mark and the following IP rules / routes:
```
$ ip rule list
...
10:     from all fwmark 0xa00/0xf00 lookup 2005

$ ip route show table 2005
default via 10.0.1.116 dev cilium_host
10.0.1.116 dev cilium_host scope link
```

One fix is to extend the troublesome check https://github.com/cilium/cilium/blob/v1.12.3/bpf/bpf_host.c#L626 by allowing proxy replies to `WORLD_ID`.

**Comment by brb** [comment]
Unfortunately, removal of the 2005 rtable is causing the legit CI failures.

After [removing the 2005 rtable to fix the L7 issue](https://github.com/cilium/cilium/pull/21980/commits), the [kube-proxy NodePort with L7 netpol](https://jenkins.cilium.io/job/cilium-master-k8s-1.24-kernel-4.9/57/testReport/Suite-k8s-1/24/K8sDatapathServicesTest_Checks_E_W_loadbalancing__ClusterIP__NodePort_from_inside_cluster__etc__with_L7_policy_Tests_NodePort_with_L7_Policy/) started to fail in the CI. After taking closer look, the removal of the rtable is causing the reply from the envoy proxy to be passed to lo instead of cilium_host :
```
14:54:33.585708 eth0  In  IP6 fc00:f853:ccd:e793::3.52394 > fc00:f853:ccd:e793::4.30239: Flags [S], seq 504540809, win 64800, options [mss 1440,sackOK,TS val 3651151592 ecr 0,nop,wscale 7], length 0
14:54:33.585852 cilium_host Out IP6 fc00:f853:ccd:e793::4.13607 > fd00:10:244:2::c527.80: Flags [S], seq 504540809, win 64800, options [mss 1440,sackOK,TS val 3651151592 ecr 0,nop,wscale 7], length 0
14:54:33.585856 cilium_net P   IP6 fc00:f853:ccd:e793::4.13607 > fd00:10:244:2::c527.80: Flags [S], seq 504540809, win 64800, options [mss 1440,sackOK,TS val 3651151592 ecr 0,nop,wscale 7], length 0
14:54:33.585916 lo    In  IP6 fd00:10:244:2::c527.80 > fc00:f853:ccd:e793::4.13607: Flags [S.], seq 2619962850, ack 504540810, win 65464, options [mss 65476,sackOK,TS val 1096880080 ecr 3651151592,nop,wscale 7], length 0
14:54:33.585960 cilium_host Out IP6 fc00:f853:ccd:e793::4.13607 > fd00:10:244:2::c527.80: Flags [R], seq 504540810, win 0, length 0
```
The NodePort request gets SNAT-ed by iptables to the cilium_host IP addr. The trace is taken on the fc00:f853:ccd:e793::4 node which runs the selected NodePort endpoint.

**Comment by brb** [comment]
x-posting https://github.com/cilium/cilium/pull/23346#issuecomment-1479755510

**Comment by brb** [comment]
Still puzzled why this doesn't happen in the v4 path:
```
07:04:13.121936 eth0  In  IP 172.12.1.6.44452 > 172.12.1.5.30452: Flags [S], seq 2923705218, win 64240, options [mss 1460,sackOK,TS val 2856811446 ecr 0,nop,wscale 7], length 0
07:04:13.122035 cilium_host Out IP 10.244.1.195.43576 > 10.244.1.237.8080: Flags [S], seq 2923705218, win 64240, options [mss 1460,sackOK,TS val 2856811446 ecr 0,nop,wscale 7], length 0
07:04:13.122038 cilium_net P   IP 10.244.1.195.43576 > 10.244.1.237.8080: Flags [S], seq 2923705218, win 64240, options [mss 1460,sackOK,TS val 2856811446 ecr 0,nop,wscale 7], length 0
07:04:13.122100 eth0  Out IP 172.12.1.5.30452 > 172.12.1.6.44452: Flags [S.], seq 1251754280, ack 2923705219, win 65483, options [mss 65495,sackOK,TS val 3449593346 ecr 2856811446,nop,wscale 7], length 0
```
The request got SNAT-ed to the `cilium_host`'s v4 addr (same as in the v6 case).

**Comment by jschwinger233** [comment]
Ok I believe I have thoroughly figured out this matter.

TL;DR

ip6tables miss some rules to allow the responding SYN+ACK  from proxy to do reverse NAT:

this is current ip6tables rules:

```
-A CILIUM_OUTPUT_raw -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: NOTRACK for proxy return traffic" -j CT --notrack
```

this is what we expect:

```
-A CILIUM_OUTPUT_raw -o lxc+ -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: NOTRACK for proxy return traffic" -j CT --notrack
-A CILIUM_OUTPUT_raw -o cilium_host -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: NOTRACK for proxy return traffic" -j CT --notrack
-A CILIUM_OUTPUT_raw -o lxc+ -m mark --mark 0x800/0xe00 -m comment --comment "cilium: NOTRACK for L7 proxy upstream traffic" -j CT --notrack
-A CILIUM_OUTPUT_raw -o cilium_host -m mark --mark 0x800/0xe00 -m comment --comment "cilium: NOTRACK for L7 proxy upstream traffic" -j CT --notrack
```

Long version:

1. Why this issue is relevant to route table 2005? 

Route table 2005 is simply routing skb with mark `0xa00` to cilium_host, and skb returned by proxy is marked by '0xa00'.

Thanks to this routing, the responding skb is going to cilium_host, then cilium_net, and finally back to kernel stack, but the mark `0xa00` is made clear after cilium_net. After that, the current `NOTRACK for proxy return traffic` ip6tables rule will be hit no more, and conntrack can rev-NAT the skb as we wish.

Once we delete route table 2005, responding skb will be routed by the following rule in local table:

```
local fd00:10:244:3::bba5 dev cilium_host proto kernel metric 0 pref medium
```

According to the kenel source code, route rules with type `local` will use `lo` if the dev is not an L3 master:

1) if route type is local, mark flags with `RTF_LOCAL`: https://github.com/torvalds/linux/blob/62bad54b26db8bc98e28749cd76b2d890edb4258/net/ipv6/route.c#L5031-L5032
2) if flags is marked with `RTF_LOCAL`, set dst dev `lo`: https://github.com/torvalds/linux/blob/62bad54b26db8bc98e28749cd76b2d890edb4258/net/ipv6/route.c#L1011-L1020

Therefore, responding skb will go to `lo` instead of `cilium_host`, and gets no chance to be rev-NAT-ed.

This explains why my PR #24208 fails to fix this issue once deleting route table 2005.

2. Why deleting route table 2005 doesn't affect ipv4?

ip4tables has rules of "NOTRACK for proxy return traffic" which are only applied for `-o cilium_host` or `-o lxc+`, so after route table 2005 disappears, the return skb which is prepared to be routed to `lo` (because of the same `local` type routing in local route table) can be processed by conntrack properly, and at last will be rerouted to eth0 and complete the tcp handshake.

This will be fixed together in #24208
